sdk/core/azure-core-amqp/src/common/global_state.cpp

```diff
@@ -138,6 +150,35 @@ namespace Azure { namespace Core { namespace Amqp { namespace Common { namespace
     }
   }
 
+  void GlobalStateHolder::RemovePollable(std::shared_ptr<Pollable> pollable)
+  {
+    // There is a bit of a complicated lock-free dance happening here.
+    // The m_pollables list is accessed by the polling thread, and the list is modified by the user
+    // thread. To ensure integrity of the list, the polling thread takes the lock, copies the
+    // pollable from the list, releases the lock and then iterates over the pollables at the
+    // snapshot.
+    //
+    // Because the pollable is a shared_ptr, the user thread can remove a pollable while the
+    // background thread is polling.
+    //
+    // But we want to make sure that the thread has finished polling (and thus has removed the copy
+    // of the pollables list). For that, we have the m_activelyPolling variable. It is set under the
+    // pollables lock, and cleared after the polling thread has finished polling (outside the lock).
+    //
+    // This means that we can spin on the m_activelyPolling variable *under* the pollables lock safe
+    // in the knowledge that IF the variable is set to true, it means that we acquired the
+    // pollablesMutex during the interval when the captured list is being interated over. And that
+    // the m_activelyPolling variable will only be cleared AFTER the captured list is freed.
+    //
+
+    std::lock_guard<std::mutex> lock(m_pollablesMutex);
+    m_pollables.remove(pollable);
+    // Spin until m_activelyPolling is false, this ensures that the polling thread is not using the
+    // capturedList copy of m_pollables.
+    while (m_activelyPolling.load())
```

Comment:
How long can we spin here? Would it make sense to sleep/use mutex instead?

Reply:
I explicitly don't want to use a mutex. This should spin for as long as it takes for the pollables to run which should be milliseconds. I can put a yield in here though if it matters.

Reply:
That sounds reasonable.
Milliseconds should be totally fine, I think. It's that if it was typically seconds, I'd suggest to sleep for 50-100msec between the spins, but even then I don't have enough context to confidently suggest that, I trust the decision that you made having all the information and knowing all the options.